pkg/verification/service/verification_service.go

```diff
@@ -322,6 +323,109 @@ func (s *ServiceImpl) VerifyPhoneCode(ctx *gin.Context, userID, username, code s
 }
 
+// VerifyActivationCode verifies the activation code:
+// - checks that the SocialEvent resource named after the activation code exists
+// - checks that the SocialEvent has enough capaticity to approve the user
+func (s *ServiceImpl) VerifyActivationCode(ctx *gin.Context, userID, username, code string) error {
+	log.Infof(ctx, "verifying activation code '%s'", code)
+	// look-up the UserSignup
+	signup, err := s.Services().SignupService().GetUserSignup(userID, username)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return crterrors.NewNotFoundError(err, "user not found")
+		}
+		return crterrors.NewInternalError(err, fmt.Sprintf("error retrieving usersignup: %s", userID))
+	}
+	annotationValues := map[string]string{}
+	annotationsToDelete := []string{}
+	unsetVerificationRequired := false
+
+	defer func() {
+		doUpdate := func() error {
+			signup, err := s.Services().SignupService().GetUserSignup(userID, username)
+			if err != nil {
+				return err
+			}
+			if unsetVerificationRequired {
+				states.SetVerificationRequired(signup, false)
+			}
+			if signup.Annotations == nil {
+				signup.Annotations = map[string]string{}
+			}
+			for k, v := range annotationValues {
+				signup.Annotations[k] = v
+			}
+			for _, annotationName := range annotationsToDelete {
+				delete(signup.Annotations, annotationName)
+			}
+			// also, label the UserSignup with the name of the SocialEvent (ie, the activation code)
+			if signup.Labels == nil {
+				signup.Labels = map[string]string{}
+			}
+			signup.Labels[toolchainv1alpha1.SocialEventUserSignupLabelKey] = code
+			_, err = s.Services().SignupService().UpdateUserSignup(signup)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}
+		if err := pollUpdateSignup(ctx, doUpdate); err != nil {
+			log.Errorf(ctx, err, "unable to update user signup after validating activation code")
+		}
+	}()
+
+	attemptsMade, err := checkAttempts(signup)
+	if err != nil {
+		return err
+	}
+	attemptsMade++
+	annotationValues[toolchainv1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+
+	// look-up the SocialEvent
+	event, err := s.CRTClient().V1Alpha1().SocialEvents().Get(code)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// The code doesn't match
+			return crterrors.NewForbiddenError("invalid code", "the provided code is invalid")
+		}
+		return crterrors.NewInternalError(err, fmt.Sprintf("error retrieving event '%s'", code))
+	}
+	// if there is room for the user and if the "time window" to signup is valid
+	now := metav1.NewTime(time.Now())
+	log.Infof(ctx, "verifying activation code '%s': event.Status.ActivationCount=%d, event.Spec.MaxAttendees=%d, event.Spec.StartTime=%s, event.Spec.EndTime=%s", code, strconv.Itoa(event.Status.ActivationCount), strconv.Itoa(event.Spec.MaxAttendees), event.Spec.StartTime.Format("2006-01-02:03:04:05"), event.Spec.EndTime.Format("2006-01-02:03:04:05"))
+
+	if event.Status.ActivationCount < event.Spec.MaxAttendees &&
```

Comment:
What updates the `ActivationCount`? Is it the host operator?

Reply:
yes, that's the `SocialEventController` which takes care of updating the status. This will be addressed in a subsequent PR (because a couple of PRs for both registration-service and host-operator at the same time would not work on OpenShift-CI)